packages/flutter_tools/lib/src/base/logger.dart

```diff
@@ -395,123 +422,154 @@ enum _LogType { error, status, trace }
 /// information when stopped. When canceled, the information isn't shown. In
 /// either case, a newline is printed.
 ///
+/// The [SummaryStatus] subclass shows only a static message (without an
+/// indicator), then updates it when the operation ends.
+///
 /// Generally, consider `logger.startProgress` instead of directly creating
 /// a [Status] or one of its subclasses.
-class Status {
-  Status({this.onFinish});
+abstract class Status {
+  Status({ @required this.timeout, this.onFinish }) : assert(timeout != null);
 
-  /// A straight [Status] or an [AnsiSpinner] (depending on whether the
+  /// A [SilentStatus] or an [AnsiSpinner] (depending on whether the
   /// terminal is fancy enough), already started.
-  factory Status.withSpinner({ VoidCallback onFinish }) {
+  factory Status.withSpinner({ @required Duration timeout, VoidCallback onFinish }) {
+    assert(timeout != null);
     if (terminal.supportsColor)
-      return AnsiSpinner(onFinish: onFinish)..start();
-    return Status(onFinish: onFinish)..start();
+      return AnsiSpinner(timeout: timeout, onFinish: onFinish)..start();
+    return SilentStatus(timeout: timeout, onFinish: onFinish)..start();
   }
 
+  final Duration timeout;
   final VoidCallback onFinish;
 
   bool _isStarted = false;
 
+  @protected
+  final Stopwatch stopwatch = Stopwatch();
+
+  @protected
+  bool get seemsSlow => stopwatch.elapsed > timeout;
+
+  @protected
+  String get elapsedTime {
+    if (timeout > fastOperation)
+      return getElapsedAsSeconds(stopwatch.elapsed);
+    return getElapsedAsMilliseconds(stopwatch.elapsed);
+  }
+
   /// Call to start spinning.
   void start() {
     assert(!_isStarted);
+    assert(!stopwatch.isRunning);
```

Comment:
If `_isStarted` and `stopwatch.isRunning` always go hand in hand, do we need `_isStarted`?

Reply:
removed _isStarted